File: LO_Rejects.R
library(tidyverse)
library(caret)

Rejects_Data <- (read.csv(file = 'Rejects_Analysis.csv', header = T))
#Remove hidden column
Rejects_Data <- select(Rejects_Data, -ï..PROCESS_ID)


library(skimr)
summaryStats <- skim(Rejects_Data)
summaryStats

#Clean missing values from DocuSign attachments validation
#Rejects_Data$DocuSign_Used <- ifelse(Rejects_Data$DocuSign_Used == "N/A", "Nothing to Sign", Rejects_Data$DocuSigned_Used)
library(lubridate)
#Clean dates
Rejects_Data <- mutate(Rejects_Data, Start_Month = month(as.Date(Rejects_Data$Start_Date, '%m/%d/%Y')),
                                     Start_Day = day(as.Date(Rejects_Data$Start_Date, '%m/%d/%Y')),
                                     Start_Year = year(as.Date(Rejects_Data$Start_Date, '%m/%d/%Y')),
                                     End_Month = month(as.Date(Rejects_Data$End_Date, '%m/%d/%Y')),
                                     End_Day = day(as.Date(Rejects_Data$End_Date, '%m/%d/%Y')),
                                     End_Year = year(as.Date(Rejects_Data$End_Date, '%m/%d/%Y')))

Rejects_Data$Start_Date <- NULL
Rejects_Data$End_Date <- NULL
Rejects_Data$End_Year <- NULL
Rejects_Data$Start_Year <- NULL
Rejects_Data$DocuSign_Used <- ifelse(Rejects_Data$DocuSign_Used == "false", "FALSE", Rejects_Data$DocuSign_Used)
Rejects_Data$DocuSign_Used <- ifelse(Rejects_Data$DocuSign_Used == "true", "TRUE", Rejects_Data$DocuSign_Used)


#Remove end dates that haven't occured yet
Rejects_Data$End_Month <- ifelse(Rejects_Data$End_Month > 3, NA, Rejects_Data$End_Month)
Rejects_Data$End_Month <- ifelse(is.na(Rejects_Data$End_Month), na.omit(Rejects_Data$End_Month), Rejects_Data$End_Month)

Rejects_Data$DaysBin <- c()
Rejects_Data$DaysBin[Rejects_Data$End_Day - Rejects_Data$Start_Day < 2] <- "Same Day"
Rejects_Data$DaysBin[Rejects_Data$End_Day - Rejects_Data$Start_Day >=2 & Rejects_Data$End_Day - Rejects_Data$Start_Day <=5] <- "2 to 5 Days"
Rejects_Data$DaysBin[Rejects_Data$End_Day - Rejects_Data$Start_Day >=6 & Rejects_Data$End_Day - Rejects_Data$Start_Day <=14] <- "Over 1 Week"
Rejects_Data$DaysBin[Rejects_Data$End_Day - Rejects_Data$Start_Day >=15 & Rejects_Data$End_Day - Rejects_Data$Start_Day <=21] <- "Over 2 Weeks"
Rejects_Data$DaysBin[Rejects_Data$End_Day - Rejects_Data$Start_Day >=22 & Rejects_Data$End_Day - Rejects_Data$Start_Day <=28] <- "Over 3 Weeks"
Rejects_Data$DaysBin[Rejects_Data$End_Month - Rejects_Data$Start_Month >= 1] <- "Over 1 Month"

#New Feature: Multiple attachments, DS used
Rejects_Data$DsAttachments <- c()
Rejects_Data$DsAttachments[Rejects_Data$Attachments > 1 & Rejects_Data$DocuSign_Used == TRUE] <- 1
Rejects_Data$DsAttachments <- ifelse(is.na(Rejects_Data$DsAttachments), 0, Rejects_Data$DsAttachments)

Rejects_Data$DS_Expiration_SameDay <- ifelse(Rejects_Data$DocuSign_Used == "TRUE" & Rejects_Data$DaysBin ==  "Same Day", 1 ,0)
Rejects_Data$DS_Expiration_Days <- ifelse(Rejects_Data$DocuSign_Used == "TRUE" & Rejects_Data$DaysBin ==  "2 to 5 Days", 1 ,0)
Rejects_Data$DS_Expiration_1wk <- ifelse(Rejects_Data$DocuSign_Used == "TRUE" & Rejects_Data$DaysBin ==  "Over 1 Week", 1 ,0)
Rejects_Data$DS_Expiration_2wks <- ifelse(Rejects_Data$DocuSign_Used == "TRUE" & Rejects_Data$DaysBin ==  "Over 2 Weeks", 1 ,0)
Rejects_Data$DS_Expiration_3wks <- ifelse(Rejects_Data$DocuSign_Used == "TRUE" & Rejects_Data$DaysBin ==  "Over 3 Weeks", 1 ,0)
Rejects_Data$DS_Expiration_1_Month <- ifelse(Rejects_Data$DocuSign_Used == "TRUE" & Rejects_Data$DaysBin ==  "Over 1 Month", 1 ,0)

Rejects_Data$Process <- ifelse(Rejects_Data$NAME == "AACR Approval Subprocess", 'AACR',
                        ifelse(Rejects_Data$NAME == "AACR Manual Update Process", "AACR", 
                        ifelse(Rejects_Data$NAME == "Account Documentation Submission Process", "ADS",
                        ifelse(Rejects_Data$NAME == "Account Retitling Process", "Acct Retitle",
                        ifelse(Rejects_Data$NAME == "ACH Instructions", "AINS", 
                        ifelse(Rejects_Data$NAME == "Advisory Goal Tool Process", "AGT",
                        ifelse(Rejects_Data$NAME == "Advisory Program Schedule Generation", "AWA",
                        ifelse(Rejects_Data$NAME == "Advisory Target Allocation Process", "ATAP",
                        ifelse(Rejects_Data$NAME == "Advisory Workflow Application (AWA)", "AWA",
                        ifelse(Rejects_Data$NAME == "Annual Advisory Client Review (AACR)", "AACR",
                        ifelse(Rejects_Data$NAME == "AWA Subsession", "AWA",
                        ifelse(Rejects_Data$NAME == "Beneficiary Change Process", "Bene Change",
                        ifelse(Rejects_Data$NAME == "Blank Form Packet Creator", "AO",
                        ifelse(Rejects_Data$NAME == "Client Transfer Application", "CTA",
                        ifelse(Rejects_Data$NAME == "CostBasis Process", "Cost Basis",
                        ifelse(Rejects_Data$NAME == "Create Bulk PDF of CRT Letters", "CRT",
                        ifelse(Rejects_Data$NAME == "DRS_DRIP Process", "DRS_DRP",
                        ifelse(Rejects_Data$NAME == "DTC (Free Delivery) Request Process", "DTC",
                        ifelse(Rejects_Data$NAME == "DWAC Subprocess", "DWAC",
                        ifelse(Rejects_Data$NAME == "Fee Waiver Process", "Fee Waiver",
                        ifelse(Rejects_Data$NAME == "FI Allocation Account Opening", "Fixed Income",
                        ifelse(Rejects_Data$NAME == "FI Master Account Opening", "Fixed Income",
                        ifelse(Rejects_Data$NAME == "Find Existing Pre-sale Disclosure", "Retirement Plans",
                        ifelse(Rejects_Data$NAME == "Household Maintenance", "HM",
                        ifelse(Rejects_Data$NAME == "IER Account Opening Process", "IER",
                        ifelse(Rejects_Data$NAME == "Insurance Submission Process", "Ins Submission",
                        ifelse(Rejects_Data$NAME == "IRA Simplifiers Process", "AO",
                        ifelse(Rejects_Data$NAME == "LiquidOffice ACAT Process", "ACAT",
                        ifelse(Rejects_Data$NAME == "Margin Option Process", "M/O",
                        ifelse(Rejects_Data$NAME == "Mutual Fund Networking Request Process", "MFNR",
                        ifelse(Rejects_Data$NAME == "Open an account", "AO",
                        ifelse(Rejects_Data$NAME == "Periodic Investment Purchase and Sell (PIPS)", "PIPS",
                        ifelse(Rejects_Data$NAME == "PIM Composite Update Process", "PIM Comp Update",
                        ifelse(Rejects_Data$NAME == "Qualified Plan Account Opening", "Retirement Plans",
                        ifelse(Rejects_Data$NAME == "Qualified Plan Maintenance", "Retirement Plans",
                        ifelse(Rejects_Data$NAME == "Qualified Plan Pre-sale Disclosure Process", "Retirement Plans",
                        ifelse(Rejects_Data$NAME == "Review Previous Day Repurchase Trades", "Review Prev Day Trades",
                        ifelse(Rejects_Data$NAME == "Schedule C Update Subprocess", "Schedule C",
                        ifelse(Rejects_Data$NAME == "SSI Approval Subprocess", "SSI",
                        ifelse(Rejects_Data$NAME == "Statement Household Maintenance", "HM",
                        ifelse(Rejects_Data$NAME == "Supervision Group Acceptance Process", "AO",
                        ifelse(Rejects_Data$NAME == "Systematic and Standing Instructions Process", "SSI",
                        ifelse(Rejects_Data$NAME == "Tax Gain or Loss Sub Process", "Tax GL",
                        ifelse(Rejects_Data$NAME == "Trade Correction Process", "Trade Corr",
                        ifelse(Rejects_Data$NAME == "Transfer to Client Resource Team Process", "TCRT",
                        ifelse(Rejects_Data$NAME == "UTMA UGMA Exception State Registration Process", "UTMA Excp",
                        ifelse(Rejects_Data$NAME == "UTMA UGMA Freeze Subprocess", "UTMA Retitle",
                        ifelse(Rejects_Data$NAME == "Verbal Account Opening Approval Subprocess", "AO", "Pls Fix"))))))))))))))))))))))))))))))))))))))))))))))))

Rejects_Data$NAME <- NULL
Rejects_Data$Start_Day <- NULL
Rejects_Data$End_Day <- NULL
Rejects_Data$Start_Month <- NULL
Rejects_Data$End_Month <- NULL

#Rejects_Data$End_Year <- ifelse(is.na(Rejects_Data$End_Year), na.omit(Rejects_Data$End_Year), Rejects_Data$End_Year)

#Dummy Variables
#get predictors without response (and those predictors with only one value as well as date/timestamp values)
Rejects_Data_predictors <- select(Rejects_Data, -Reject)
#create dummy vars expect for the response
dummies_model <- dummyVars(~., data = Rejects_Data_predictors)

#provide only predictors that are now convert to dummy variables
Rejects_Data_predictors_dummy <- data.frame(predict(dummies_model, newdata = Rejects_Data))

#recombine predictors including dummy variables with response
Rejects_Data <- cbind(Reject = Rejects_Data$Reject, Rejects_Data_predictors_dummy)

#Drop NA's (1 for each bin)
Rejects_Data$DaysBin2.to.5.Days <- ifelse(is.na(Rejects_Data$DaysBin2.to.5.Days), na.omit(Rejects_Data$DaysBin2.to.5.Days), Rejects_Data$DaysBin2.to.5.Days)
Rejects_Data$DaysBinOver.1.Month <- ifelse(is.na(Rejects_Data$DaysBinOver.1.Month), na.omit(Rejects_Data$DaysBinOver.1.Month), Rejects_Data$DaysBinOver.1.Month)
Rejects_Data$DaysBinOver.1.Week <- ifelse(is.na(Rejects_Data$DaysBinOver.1.Week), na.omit(Rejects_Data$DaysBinOver.1.Week), Rejects_Data$DaysBinOver.1.Week)
Rejects_Data$DaysBinOver.2.Weeks <- ifelse(is.na(Rejects_Data$DaysBinOver.2.Weeks), na.omit(Rejects_Data$DaysBinOver.2.Weeks), Rejects_Data$DaysBinOver.2.Weeks)
Rejects_Data$DaysBinOver.3.Weeks <- ifelse(is.na(Rejects_Data$DaysBinOver.3.Weeks), na.omit(Rejects_Data$DaysBinOver.3.Weeks), Rejects_Data$DaysBinOver.3.Weeks)
Rejects_Data$DaysBinSame.Day <- ifelse(is.na(Rejects_Data$DaysBinSame.Day), na.omit(Rejects_Data$DaysBinSame.Day), Rejects_Data$DaysBinSame.Day)

#Convert FRAUD_NOFRAUD to a factor
Rejects_Data$Reject <- as.factor(Rejects_Data$Reject)
Rejects_Data$Reject <- fct_recode(Rejects_Data$Reject, NoReject = 'No_Reject', Reject = 'Reject')

#Logistic Classification
#set.seed()
#Split data for partitioning and indexing 
set.seed(99)
index <- createDataPartition(Rejects_Data$Reject,p = .8, list = FALSE)
Rejects_data_train <- Rejects_Data[index,]
Rejects_data_test <- Rejects_Data[-index,]

set.seed(99)



# Rejects_LOG_CLASS_model <- train(Reject~ ., 
#                                data = Rejects_data_train,
#                                method = 'glmnet',
#                                #method = 'cv' cv is cross validation
#                                trControl = trainControl(method = 'cv',
#                                                         #number lets the model know how many validation sets to create
#                                                         number = 5,
#                                                         classProbs = TRUE,
#                                                         summaryFunction = twoClassSummary),
#                                metric = 'ROC')
# Rejects_LOG_CLASS_model
# #Variable importance
# plot(varImp(Rejects_LOG_CLASS_model))
# varImp(Rejects_LOG_CLASS_model)
# #List coefficients selected
# coef(Rejects_LOG_CLASS_model$finalModel, Rejects_data_test$bestTune$lambda)
# #Get the predicted probabilities of the test data
# predprob_lasso_LOG <- predict(Rejects_LOG_CLASS_model, Rejects_data_test, type = 'prob')
# #Use cbind() to bring together probs and tran_id
# #Builds confusion matrix
# library(ROCR)
# pred_lasso_log <- prediction(predprob_lasso_LOG$Reject, Rejects_data_test$Reject, label.ordering = c("NoReject", "Reject"))
# 
# #tpr and fpr = True positive rate and false positive rate
# perf_lasso_log <- performance(pred_lasso_log, "tpr", "fpr")
# plot(perf_lasso_log, colorize = TRUE)
# 
# #Get the AUC
# auc_lasso_log <- unlist(slot(performance(pred_lasso_log, "auc"), "y.values"))
# auc_lasso_log

#Classification Decision Tree
#Split data at 8:2
#install.packages('rpart')
library(rpart)
set.seed(99)
Rejects_DTree_model <- train(Reject~.,
                           data = Rejects_data_train,
                           method = 'rpart',
                           trControl = trainControl(method = 'cv', number = 5,
                                                    #Estimate class probabilities
                                                    classProbs = TRUE,
                                                    #needed to get ROC
                                                    summaryFunction = twoClassSummary),
                           metric = 'ROC')
#Prunning to avoid overfitting

#Provides information of parameter tuning via cross validation
Rejects_DTree_model

plot(Rejects_DTree_model) #provides plot of paraneter tuning via cross validation

#plot variable importance
plot(varImp(Rejects_DTree_model))

#install.packages("rpart.plot")
library(rpart.plot)
rpart.plot(Rejects_DTree_model$finalModel, type = 5)

#First, get the prdicted probablities of the test data.
predprob_rejects_DTree <- predict(Rejects_DTree_model, Rejects_data_test, type = 'prob')

library(ROCR)
pred_lasso_DTree <- prediction(predprob_rejects_DTree[,2], Rejects_data_test$Reject, label.ordering = c('NoReject', 'Reject'))
perf_lasso_DTree <- performance(pred_lasso_DTree, 'tpr', 'fpr')
plot(perf_lasso_DTree, colorize = TRUE)

#Get AUC
auc_lasso_DTree <- unlist(slot(performance(pred_lasso_DTree, "auc"), "y.values"))

auc_lasso_DTree

#XGBoost
library(xgboost)
set.seed(99)
Rejects_xgb_model <- train(Reject~.,
                         data = Rejects_data_train,
                         method = 'xgbTree',
                         trControl = trainControl(method = 'cv', number = 5,
                                                  #Estimate class probabilities
                                                  classProbs = TRUE,
                                                  #needed to get ROC
                                                  summaryFunction = twoClassSummary),
                         metric = 'ROC')
#Prunning to avoid overfitting

Rejects_xgb_model
Rejects_xgb_model$bestTune

plot(Rejects_xgb_model) #provides plot of parameter tuning via cross validation

#plot variable importance
plot(varImp(Rejects_xgb_model))
varImp(Rejects_xgb_model)

#SHAP for interactions of variables. Wasn't a lot to see here. Earlier scatterplots were much more helpful
#install.packages('SHAPforxgboost')
library(SHAPforxgboost)
xdata <- as.matrix(select(Rejects_data_train, -Reject))
shap <- shap.prep(Rejects_xgb_model$finalModel, X_train = xdata)
shap.plot.summary(shap)
top4 <- shap.importance(shap, names_only = TRUE)[1:4]
for (x in top4) {
  p <- shap.plot.dependence(
    shap, 
    x = x, 
    color_feature = "auto", 
    smooth = FALSE, 
    jitter_width = 0.01, 
    alpha = 0.4
  ) +
    ggtitle(x)
  print(p)
}

#First, get the prdicted probablities of the test data.
predprob_rejects_xgb <- predict(Rejects_xgb_model, Rejects_data_test, type = 'prob')
#use cbind() to bring together predprob[,2] with tran_id

library(ROCR)
pred_lasso_xgb <- prediction(predprob_rejects_xgb[,2], Rejects_data_test$Reject, label.ordering = c('NoReject', 'Reject'))
perf_lasso_xgb <- performance(pred_lasso_xgb, 'tpr', 'fpr')
plot(perf_lasso_xgb, colorize = TRUE)

#Get AUC
auc_lasso_xgb <- unlist(slot(performance(pred_lasso_xgb, "auc"), "y.values"))

auc_lasso_xgb

#Random Forest
library(randomForest)
rf_grid <- expand.grid(mtry = c(1, 3, 7))
rf_grid

set.seed(99)
Rejects_rf_model <- train(Reject~.,
                        data = Rejects_data_train,
                        method = "rf",
                        trControl = trainControl(method = "cv", number = 5,
                                                 classProbs = TRUE,
                                                 summaryFunction = twoClassSummary),
                        metric = "ROC",
                        tuneGrid = rf_grid,
                        
                        verbose = FALSE)
Rejects_rf_model

plot(Rejects_rf_model)

plot(varImp(Rejects_rf_model))

predprob_rejects_rf <- predict(Rejects_rf_model, Rejects_data_test, type = 'prob')
#use cbind() to bring together predprob[,2] with tran_id

library(ROCR)
pred_lasso_rf <- prediction(predprob_rejects_rf[,2], Rejects_data_test$Reject, label.ordering = c('NoReject', 'Reject'))
perf_lasso_rf <- performance(pred_lasso_rf, 'tpr', 'fpr')
plot(perf_lasso_rf, colorize = TRUE)

auc_lasso_rf <- unlist(slot(performance(pred_lasso_rf, "auc"), "y.values"))

auc_lasso_rf

File: LO_Rejects.py
from asyncore import write
from cProfile import label
from datetime import datetime
from operator import index
from pickle import TRUE
from tabnanny import verbose
from tkinter import Grid, Variable
from turtle import tiltangle
from typing import Sequence
from unittest import result
import matplotlib
from matplotlib import dates
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from datetime import datetime, date, time as dt
from skimpy import skim

Rejects_imp = 'Rejects_Analysis.csv'

Rejects = pd.read_csv(Rejects_imp)
skim(Rejects)
Rejects['DocuSigned_Used'] = np.where(Rejects['DocuSign_Used'] == 'TRUE', 1, 0)
Rejects['Reject'] = np.where(Rejects['Reject'] == 'Reject', 1, 0)
Rejects['DocuSign_Used'] = np.where(Rejects['DocuSign_Used'] == 'TRUE', 1, 0)

Rejects['DS_Expiration_SameDay'] = np.logical_and(Rejects['DocuSign_Used'] == 1, Rejects['Same Day'] == 1)
Rejects['DS_Expiration_SameDay'] = np.where((Rejects['DS_Expiration_SameDay'] == True), 1, 0)
Rejects['DS_Expiration_Days'] = np.logical_and(Rejects['DocuSign_Used'] == 1, Rejects['2 to 5 Days'] == 1)
Rejects['DS_Expiration_Days'] = np.where((Rejects['DS_Expiration_Days'] == True), 1, 0)
Rejects['DS_Expiration_1wk'] = np.logical_and(Rejects['DocuSign_Used'] == 1, Rejects['Over 1 Week'] == 1)
Rejects['DS_Expiration_1wk'] = np.where((Rejects['DS_Expiration_1wk'] == True), 1, 0)
Rejects['DS_Expiration_2wks'] = np.logical_and(Rejects['DocuSign_Used'] == 1, Rejects['Over 2 Weeks'] == 1)
Rejects['DS_Expiration_2wks'] = np.where((Rejects['DS_Expiration_2wks'] == True), 1, 0)
Rejects['DS_Expiration_3wks'] = np.logical_and(Rejects['DocuSign_Used'] == 1, Rejects['Over 3 Weeks'] == 1)
Rejects['DS_Expiration_3wks'] = np.where((Rejects['DS_Expiration_3wks'] == True), 1, 0)
Rejects['DS_Expiration_Month'] = np.logical_and(Rejects['DocuSign_Used'] == 1, Rejects['Over 1 Month'] == 1)
Rejects['DS_Expiration_Month'] = np.where((Rejects['DS_Expiration_Month'] == True), 1, 0)

Rejects['NAME'] = np.where((Rejects['NAME'] == 'AACR Approval Subprocess'), 'AACR', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'AACR Manual Update Process'), 'AACR', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Account Documentation Submission Process'), 'ADS', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Account Retitling Process'), 'Acct Retitle', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'ACH Instructions'), 'AINS', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Advisory Goal Tool Process'), 'AGT', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Advisory Program Schedule Generation'), 'AWA', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Advisory Target Allocation Process'), 'ATAP', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Advisory Workflow Application (AWA)'), 'AWA', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Annual Advisory Client Review (AACR)'), 'AACR', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'AWA Subsession'), 'AWA', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Beneficiary Change Process'), 'Bene Change', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Blank Form Packet Creator'), 'AO', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Client Transfer Application'), 'CTA', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'CostBasis Process'), 'Cost Basis', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Create Bulk PDF of CRT Letters'), 'CRT', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'DRS_DRIP Process'), 'DRS_DRP', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'DTC (Free Delivery) Request Process'), 'DTC', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'DWAC Subprocess'), 'DWAC', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Fee Waiver Process'), 'Fee Waiver', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'FI Allocation Account Opening'), 'Fixed Income', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'FI Master Account Opening'), 'Fixed Income', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Find Existing Pre-sale Disclosure'), 'Retirement Plans', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Household Maintenance'), 'HM', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'IER Account Opening Process'), 'IER', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Insurance Submission Process'), 'Ins Submission', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'IRA Simplifiers Process'), 'AO', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'LiquidOffice ACAT Process'), 'ACAT', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Margin Option Process'), 'M/O', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Mutual Fund Networking Request Process'), 'MFNR', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Open an account'), 'AO', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Periodic Investment Purchase and Sell (PIPS)'), 'PIPS', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'PIM Composite Update Process'), 'PIM Comp Update', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Qualified Plan Account Opening'), 'Retirement Plans', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Qualified Plan Maintenance'), 'Retirement Plans', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Qualified Plan Pre-sale Disclosure Process'), 'Retirement Plans', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Review Previous Day Repurchase Trades'), 'Review Prev Day Trades', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Schedule C Update Subprocess'), 'Schedule C', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'SSI Approval Subprocess'), 'SSI', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Statement Household Maintenance'), 'HM', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Supervision Group Acceptance Process'), 'AO', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Systematic and Standing Instructions Process'), 'SSI', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Tax Gain or Loss Sub Process'), 'Tax GL', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Trade Correction Process'), 'Trade Corr', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Transfer to Client Resource Team Process'), 'TCRT', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'UTMA UGMA Exception State Registration Process'), 'UTMA Excep', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'UTMA UGMA Freeze Subprocess'), 'UTMA Retitle', Rejects['NAME'])
Rejects['NAME'] = np.where((Rejects['NAME'] == 'Verbal Account Opening Approval Subprocess'), 'AO', Rejects['NAME'])

df = Rejects['NAME'].str.get_dummies()
Rejects = pd.concat([Rejects, df], axis=1)


Rejects_pid = Rejects.drop(['PROCESS_ID', 'NAME', 'Start_Date', 'End_Date', 'Start_Day', 'Start_Month', 'End_Day', 'End_Month', 'DocuSigned_Used'], axis=1)
Rejects = Rejects_pid

skim(Rejects)



#Split into train, validation, and test sets
from sklearn.model_selection import train_test_split

features = Rejects.drop('Reject', axis=1)
labels = Rejects['Reject']

X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=.2, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_test, y_test, test_size=.5, random_state=42)

for dataset in [y_train, y_val, y_test]:
    print(round(len(dataset)/ len(labels), 2))

X_train.to_csv('rejects_train_features.csv', index = False)
X_val.to_csv('val_features.csv', index = False)
X_test.to_csv('test_features.csv', index = False)

y_train.to_csv('rejects_train_labels.csv', index = False)
y_val.to_csv('val_labels.csv', index = False)
y_test.to_csv('test_lables.csv', index = False)


import joblib
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
#Grid Seach assist in hyperparameter tuning
import warnings
warnings.filterwarnings('ignore', category=FutureWarning)
warnings.filterwarnings('ignore', category=DeprecationWarning)

tr_features = pd.read_csv('rejects_train_features.csv')
tr_labels = pd.read_csv('rejects_train_labels.csv')

#Create function to neatly present best parameteres when testing models
def print_results(results):
    print('Best Params: {}\n'.format(results.best_params_))

    means = results.cv_results_['mean_test_score']
    stds = results.cv_results_['std_test_score']
    for mean, std, params in zip(means, stds, results.cv_results_['params']):
        print('{} (+/-{}) for {}'.format(round(mean, 3), round(std*2, 3), params))

Rejects.head()

#Logistic Regression
#When being able to explain how variables affect model is important. 

lr = LogisticRegression()
parameters = {
    'C': [.001, .01, .1, 1, 10, 100, 1000]
}

cv = GridSearchCV(lr, parameters,  cv=5)
cv.fit(tr_features, tr_labels.values.ravel())

print_results(cv)

#Pickel and Save model to compare for later
joblib.dump(cv.best_estimator_,'LR_model.pkl')


#Support Vector Machine (SVC)
#Data is short and fat. Many variables but not a lot of rows/observations
from sklearn.svm import SVC
svc = SVC()
parameters = {
    'kernel': ['linear', 'rbf'],
    'C': [.1, 1, 10]
}

cv = GridSearchCV(svc, parameters, cv=5)
cv.fit(tr_features, tr_labels.values.ravel())

print_results(cv)
#Pickel and Save model to compare for later
joblib.dump(cv.best_estimator_,'SVM_model.pkl')

#Multilayer Perceptron (MLP)
#Complex....
from sklearn.neural_network import MLPClassifier
mlp = MLPClassifier()
parameters = {
    'hidden_layer_sizes': [(10,), (50,), (100,)],
    'activation': ['relu', 'tanh', 'logistic'],
    'learning_rate': ['constant', 'invscaling', 'adaptive']
}

cv = GridSearchCV(mlp, parameters, cv=5)
cv.fit(tr_features, tr_labels.values.ravel())

print_results(cv)

#Pickel and save model
joblib.dump(cv.best_estimator_,'MLP_model.pkl')

#Random Forest
#Few but Deep trees. Complex interactions and relationship
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier()
parameters = {
    'n_estimators': [5, 50, 250],
    'max_depth': [2, 4, 8, 16, 32, None]
}

cv = GridSearchCV(rf, parameters, cv=5)
cv.fit(tr_features, tr_labels.values.ravel())

print_results(cv)
#Pickel model and save. 
joblib.dump(cv.best_estimator_, 'RF_model.pkl')

#Gradient Boosting
#Many, but shallow trees. Complex interactions and relationships
from sklearn.ensemble import GradientBoostingClassifier
gb = GradientBoostingClassifier()
parameters = {
    'n_estimators': [5, 50, 250, 500],
    'max_depth': [1, 3, 5, 7, 9],
    'learning_rate': [0.01, 0.1, 1, 10, 100]
}

cv = GridSearchCV(gb, parameters, cv=5)

cv.fit(tr_features, tr_labels.values.ravel())

print_results(cv)

#Pickel and save model
joblib.dump(cv.best_estimator_, 'GB_model.pkl')

#Neural Network Sequential

from numpy.random import seed
seed(99)
import tensorflow
tensorflow.random.set_seed(99)
from keras.models import Sequential
from keras.layers import Dense
Nnet = Sequential()
number_inputs = 48
number_hidden_nodes = 7
Nnet.add(Dense(units=number_inputs, activation='sigmoid', input_dim = number_inputs))
Nnet.add(Dense(units=3, activation='relu'))
Nnet.add(Dense(units=2, activation='tanh'))
Nnet.add(Dense(units=1, activation='sigmoid'))
Nnet.summary()
Rejects.shape
Nnet.compile(optimizer='adam',
            loss='binary_crossentropy',
            metrics=['accuracy'])

Nnet.fit(tr_features, tr_labels, epochs=1000, shuffle=True, verbose=2)

#Evaluate Models

from sklearn.metrics import accuracy_score, precision_score, recall_score 
from time import strptime, time

val_features = pd.read_csv('val_features.csv')
val_labels = pd.read_csv('val_labels.csv')

te_features = pd.read_csv('test_features.csv')
te_labels = pd.read_csv('test_lables.csv')

models = {}

for mdl in ['LR', 'SVM', 'MLP', 'RF', 'GB']:
    models[mdl] = joblib.load('{}_model.pkl'.format(mdl))

def evaluate_model(name, model, features, labels):
    start = time()
    pred = model.predict(features)
    end = time()
    accuracy = round(accuracy_score(labels, pred), 3)
    precision = round(precision_score(labels, pred), 3)
    recall = round(recall_score(labels, pred), 3)
    print('{} -- Accuracy: {} / Precision: {} / Recall: {} / Latency: {}ms'.format(name,
                                                                                    accuracy,
                                                                                    precision,
                                                                                    recall,
                                                                                    round((end - start) * 1000, 1)))

for name, mdl in models.items():
    evaluate_model(name, mdl, val_features, val_labels)

Nnet_test_loss, NNet_test_accuracy = Nnet.evaluate(te_features, te_labels, verbose=2)

#Evaluate model that performed best on training set, on the holdout test set
evaluate_model('Random Forest', models['RF'], te_features, te_labels)
evaluate_model('Gradient Boosting', models['GB'], te_features, te_labels)
evaluate_model('SVM', models['SVM'], te_features, te_labels)
evaluate_model('MLP', models['MLP'], te_features, te_labels)
evaluate_model('Logistic Regr', models['LR'], te_features, te_labels)
print('NNet Loss {}: NNet Accuracy {}'.format(round(Nnet_test_loss, 2), round(NNet_test_accuracy, 3)))
